### ISSUE TITLE
core: BlockFetcher.GetBlockByHeight is susceptible to a panic due to insufficient nil checks

### ISSUE THREAD

**Issue by odeke-em** [celestiaorg/celestia-node#3816]
If we examine this code https://github.com/celestiaorg/celestia-node/blob/896a46b5774c8b6dd26c0b071776c45add92a3f3/core/fetcher.go#L76-L81 we notice that the first condition checks `if res != nil and then if res.Block == nil`. However by the time that we are checking `if res != nil` it can really be nil, yet on line 81 we are nakedly dereferncing res.Block

## Suggestion
That code could be made much more whole and include a check for `if res == nil || res.Block == nil {` and then also add a test.

/cc @Wondertan @liamsi @cristaloleg

**Comment by Wondertan** [comment]
That's true, although in practise this case is not happening. Generally we think this result based api is confusing is error prone.

Care for a PR?
